Feature/slonik plugin

megquirk opened the pull request:
### Summary

- Moved [Slonik Plugin](https://github.com/autotelic/harpocrates/tree/main/plugins/fastify-slonik) from Harpocrates repo.

### Test Plan

- Clone and install `npm i`
- To confirm the plugin is working, clone and install Harpocrates project from [this repo](https://github.com/autotelic/harpocrates)
  - `cd` into the project directory and edit location of fastifySlonik in app.js
  ```js
  ...
  const fastifySlonik = require('<path to this plugin, i.e. ../fastify-slonik>')
  ...
  ```
  - Create .envrc `cp .envrc.example .envrc`
  - Update .envrc with public ip
    ```sh
    $ ifconfig -u | grep 'inet ' | grep -v 127.0.0.1 | cut -d\  -f2 | head -1
    # <your-public-ip> <-- Copy this
    ```
  - Update .envrc with copied public-ip `export PUBLIC_IP=<your-public-ip>`
  - Run the following commands to create and run [docker](https://www.docker.com/) container `docker-compose up`
  - Seed and migrate database
  ```sh
  $ cp test_seeds/1605303453216_seeds.js migrations/
  $ npm run migrate up
  ```
  - Run in development `npm run dev`
- If plugin is working correctly, GET request to `localhost:3000/users/1` should return
```js
{
userId: 1,
githubUsername: "Ed",
emailAddress: "Ed@autotelic.com",
firstLogin: 1610521099724,
mostRecentLogin: 1610521099724
}
```

### To Do
- index.test.js is failing, need to create example.js instead of requiring harpocrates app to test

megquirk committed:
Add package-lock.json

megquirk committed:
Update dependencies and scripts

megquirk committed:
Rename test file

megquirk committed:
Import code from @autotelic/harpocrates/plugins/fastify-slonik

megquirk committed:
Remove example

megquirk committed:
docs: Add summary for plugin

megquirk committed:
Add newlines at then end of files

eadmundo commented:
This is looking great! What I would do to make it easier to review & test is to create an accompanying PR in the harpocrates repo which uses this plugin - you can use `autotelic/fastify-slonik#feature/slonik-plugin` as the package name in the dependencies. That PR can then be updated once we release this to npm.

jkirkpatrick24 reviewed (approved):
LGTM - Lets just remove the empty test.js file 👍 Great stuff!

megquirk committed:
docs: Update to include test plan

megquirk committed:
Remove test.js

jkirkpatrick24 commented on index.js:
```diff
@@ -1,13 +1,27 @@
+  fastify.decorate('sql', sql)
+}
+
+module.exports = fastifyPlugin(fastifySlonik)
```
With fresh eyes, I realize we should have a name for this plugin. See the [fastify docs here](https://github.com/fastify/fastify-plugin#name)

The TLDR is it allows us to do dependency checks, when loading other plugins. (Make sure this plugin exists and has been registered before we try to use it)

megquirk committed:
docs: Remove coupled harpocrates test

megquirk committed:
docs: Add usage example

megquirk committed:
Add plugin options

megquirk committed:
Update test script path

megquirk committed:
Create helper variables for testing

megquirk committed:
Import tests from harpocrates repo and update with exports from helpers.js

megquirk committed:
fix: remove incorrect dependencies

megquirk committed:
Add example database variable

megquirk committed:
docs: Update to include slonik API and .envrc instructions

megquirk committed:
Remove extra lines

megquirk commented:
I think I've taken care of most of the issues:
- README updated
- tests are all passing
- plugin has been named

jkirkpatrick24 reviewed (commented):
This is looking really really good. Docs are great and the tests cover what we want. I think the last piece is to setup a postgres instance with docker and to have the tests connect to that docker instance.

We can cargo-cult a lot from `fastify-postgres`

[running postgres on docker](https://github.com/fastify/fastify-postgres/blob/master/package.json#L13)

This will allow anyone to run the tests successfully, as well as allow us to configure tests to run in a CI environment (which we are going to want for sure)

matthew-charles-chan commented:
> This is looking really really good. Docs are great and the tests cover what we want. I think the last piece is to setup a postgres instance with docker and to have the tests connect to that docker instance.
> 
> We can cargo-cult a lot from `fastify-postgres`
> 
> [running postgres on docker](https://github.com/fastify/fastify-postgres/blob/master/package.json#L13)
> 
> This will allow anyone to run the tests successfully, as well as allow us to configure tests to run in a CI environment (which we are going to want for sure)

We can take reference from the `fastify-postgres` [Development and Testing docs](https://github.com/fastify/fastify-postgres#development-and-testing) to add to our docs as well

megquirk committed:
docs: Update to include docker and postgres testing

megquirk committed:
Update scripts to include postgres

megquirk committed:
Update variables accordingly to match scripts

megquirk commented:
Updated 

> This is looking really really good. Docs are great and the tests cover what we want. I think the last piece is to setup a postgres instance with docker and to have the tests connect to that docker instance.
> 
> We can cargo-cult a lot from `fastify-postgres`
> 
> [running postgres on docker](https://github.com/fastify/fastify-postgres/blob/master/package.json#L13)
> 
> This will allow anyone to run the tests successfully, as well as allow us to configure tests to run in a CI environment (which we are going to want for sure)



> > This is looking really really good. Docs are great and the tests cover what we want. I think the last piece is to setup a postgres instance with docker and to have the tests connect to that docker instance.
> > We can cargo-cult a lot from `fastify-postgres`
> > [running postgres on docker](https://github.com/fastify/fastify-postgres/blob/master/package.json#L13)
> > This will allow anyone to run the tests successfully, as well as allow us to configure tests to run in a CI environment (which we are going to want for sure)
> 
> We can take reference from the `fastify-postgres` [Development and Testing docs](https://github.com/fastify/fastify-postgres#development-and-testing) to add to our docs as well

I've updated the scripts and docs to setup  a postgres instance with docker.

megquirk committed:
build: Add auto release action

megquirk committed:
Move release action to separate branch

jkirkpatrick24 reviewed (approved):
LGTM! Great  job. 

There is just one empty pesky test.js file we should remove but then squash and merge away!
